src/H5FDint.c

```diff
@@ -599,7 +602,517 @@ H5FD_write_vector(H5FD_t *file, uint32_t count, H5FD_mem_t types[], haddr_t addr
+    HDassert((count == 0) || (element_sizes[0] != 0));
+
+    /* Get proper DXPL for I/O */
+    dxpl_id = H5CX_get_dxpl();
+
+#ifndef H5_HAVE_PARALLEL
+    /* The no-op case
+     *
+     * Do not return early for Parallel mode since the I/O could be a
+     * collective transfer.
+     */
+    if (0 == count) {
+        HGOTO_DONE(SUCCEED)
+    }
+#endif /* H5_HAVE_PARALLEL */
+
+    if (file->base_addr > 0) {
+
+        /* apply the base_addr offset to the offsets array.  Must undo before
+         * we return.
+         */
+        for (i = 0; i < count; i++) {
+
+            offsets[i] += file->base_addr;
+        }
+        offsets_cooked = TRUE;
+    }
+
+    /* If the file is open for SWMR read access, allow access to data past
+     * the end of the allocated space (the 'eoa').  This is done because the
+     * eoa stored in the file's superblock might be out of sync with the
+     * objects being written within the file by the application performing
+     * SWMR write operations.
+     */
+    /* For now at least, only check that the offset is not past the eoa, since
+     * looking into the highest offset in the selection (different from the
+     * bounds) is potentially expensive.
+     */
+    if (!(file->access_flags & H5F_ACC_SWMR_READ)) {
+        haddr_t eoa;
+
+        if (HADDR_UNDEF == (eoa = (file->cls->get_eoa)(file, type)))
+            HGOTO_ERROR(H5E_VFL, H5E_CANTINIT, FAIL, "driver get_eoa request failed")
+
+        for (i = 0; i < count; i++) {
+
+            if ((offsets[i]) > eoa)
+
+                HGOTO_ERROR(H5E_ARGS, H5E_OVERFLOW, FAIL, "addr overflow, offsets[%d] = %llu, eoa = %llu",
+                            (int)i, (unsigned long long)(offsets[i]), (unsigned long long)eoa)
+        }
+    }
+
+    /* if the underlying VFD supports selection read, make the call */
+    /*if (file->cls->read_selection) {
+
+        if ((file->cls->read_selection)(file, count, type, mem_spaces, file_spaces, offsets, element_sizes,
+    bufs) < 0)
+
+            HGOTO_ERROR(H5E_VFL, H5E_READERROR, FAIL, "driver read selection request failed")
+    }
+    else*/
+    {
+        hsize_t        file_off[H5FD_SEQ_LIST_LEN];
+        size_t         file_len[H5FD_SEQ_LIST_LEN];
+        hsize_t        mem_off[H5FD_SEQ_LIST_LEN];
+        size_t         mem_len[H5FD_SEQ_LIST_LEN];
+        size_t         file_seq_i;
+        size_t         mem_seq_i;
+        size_t         file_nseq;
+        size_t         mem_nseq;
+        size_t         io_len;
+        size_t         dummy_nelem;
+        H5S_sel_iter_t file_iter;
+        H5S_sel_iter_t mem_iter;
+
+        /* otherwise, implement the selection read as a sequence of regular
+         * read calls.
+         */
+        extend_sizes = FALSE;
+
+        for (i = 0; i < count; i++) {
+
+            /* we have already verified that sizes[0] != 0 and
+             * types[0] != H5FD_MEM_NOLIST
+             */
```

Comment:
I see you are extending the sizes[] / types[] optimization to selection I/O proper.

Given that selections are a much heavier weight data structure, is this optimization significant?  To put it another way, is it buying us enough in terms of reduced footprint / ease of use to make up for the extra overhead?

Your call on this, but if you decide to keep it, we need to update the RFC accordingly, and also update the above comment to refer to just the element_sizes[] array.

Reply:
I view it as mostly a convenience item since in most cases the element sizes will be the same. I don't think it adds much overhead (it should be faster if all the element sizes are the same) and I think it makes sense to use a similar interface to that for vector I/O. There's currently no optimization for types since there's only one type in selection I/O, I will fix that comment. I am still unsure about not having a types array though (again, I think it should be similar to the vector I/O interface).

Reply:
The overhead I was referring to was added development complexity.  If nothing else, the optimization makes the vector sort routine a bit more painful.  Similarly it makes adding support for vector I/O a little harder in the VFD proper.  In the case of vector I/O representation of point selections, there can be a significant reduction in memory foot print which justifies the optimization in my mind -- I'm not so sure about selection I/O.

That said, you make a reasonable case.  If you are happy with it, I will add it to the RFC.

As to selection I/O not having a types vector -- if you can think of a plausible use case, I'll put it in the RFC.